File: CPP_PART/testing/time/test/ShortestPathDijkstra.cpp
//
// Created by vm on 03.11.17.
//

#include "ShortestPathDijkstra.h"

ShortestPathDijkstra::ShortestPathDijkstra(const std::string &fileName, int iterations) : Algorithm(fileName,
                                                                                                    iterations), g(Graph(v)) {}

void ShortestPathDijkstra::prepareTestData() {
        //Kraków
    g.addEdge(0,1,168);
    g.addEdge(0,5,116);//kielce
    g.addEdge(0,2,81); //katowice

    // Rzeszów
    g.addEdge(1,0,168); //Krkaów
    g.addEdge(1,5,157); // Kielce
    g.addEdge(1,6,175); // Lublin

    // Katowice
    g.addEdge(2,0,81); // Kraków
    g.addEdge(2,4, 75); // Czestochowa
    g.addEdge(2,3, 199); // Wrcoław

    // Wrocław
    g.addEdge(3,2,199); // Katowice
    g.addEdge(3,4,208); // Czestochowa
    g.addEdge(3,9,222); // łódz
    g.addEdge(3,8,186); // Poznań
    g.addEdge(3,7,210); // Zielona Góra

    //Czestochowa
    g.addEdge(4,2,75); // Katowice
    g.addEdge(4,3,208); // Wrocław
    g.addEdge(4,5,128); // Kielce
    g.addEdge(4,9,126); // Lodź

    // Kielce
    g.addEdge(5,4,128); // Częstochowa
    g.addEdge(5,0,116); // Kraków
    g.addEdge(5,1,157); // Rzeszów
    g.addEdge(5,6,184); // Lublin
    g.addEdge(5,10,180); // Warszawa
    g.addEdge(5,9,159); // Lódź

    // Lublin
    g.addEdge(6,1,175); // Rzeszów
    g.addEdge(6,5,184); // Kielce
    g.addEdge(6,10,170); // Warszawa
    g.addEdge(6,11,247); // Białystok

    //Zielona Góra
    g.addEdge(7,3,210); // Wrocław
    g.addEdge(7,8,151); // Poznań
    g.addEdge(7,13,112); // GorzówWlkp

    // Pożnań
    g.addEdge(8,7,151); // Zielona Góra
    g.addEdge(8,3,186); // Wrocław
    g.addEdge(8,9,205); // Lóddz
    g.addEdge(8,13,162); // Gorzów Wlkp
    g.addEdge(8,12,167); // Toruń

    // Lódź
    g.addEdge(9,3,222); // Wrocław
    g.addEdge(9,4,126); // Częstochowa
    g.addEdge(9,5,159); // Kielca
    g.addEdge(9,10,136); // Warszawa
    g.addEdge(9,15,297); // Olsztyn
    g.addEdge(9,12,181); // Toruń

    // Warszawa
    g.addEdge(10,6,170); // Lublin
    g.addEdge(10,5,180); // Kielce
    g.addEdge(10,9,136); // Lodz
    g.addEdge(10,11,203); // Bialystok
    g.addEdge(10,16,295); // Suwałki
    g.addEdge(10,15,216); // Olsztyn

    // Białystok
    g.addEdge(11,6,247); // Lublin
    g.addEdge(11,10,203); // Warszawa
    g.addEdge(11,15,226); // Olsztyn
    g.addEdge(11,16,129); // Suwałki

    // Toruń
    g.addEdge(12,9,181); // łódz
    g.addEdge(12,8,167); //  Poznań
    g.addEdge(12,15,176); // Olsztyn
    g.addEdge(12,19,170); // Gdansk
    g.addEdge(12,14,143); // Pila

    // Gorzów Wlkp
    g.addEdge(13,7,112); // Zielona Góra
    g.addEdge(13,8,162); // Poznan
    g.addEdge(13,17,111); // Szczecin
    g.addEdge(13,14,121); // Pila

    // Piła
    g.addEdge(14,12,143); // Toruń
    g.addEdge(14,13,121); // Gorzów Wlkp
    g.addEdge(14,19,223); // Gdańsk
    g.addEdge(14,18,140); // Koszalin

    // Olsztyn
    g.addEdge(15,16,196); // Suwałki
    g.addEdge(15,11,226); // Białystok
    g.addEdge(15,10,216); // Warszawa
    g.addEdge(15,9,297); // łodz
    g.addEdge(15,12,176); // Torun
    g.addEdge(15,19,168); // Gdansk

    // Suwałki
    g.addEdge(16,11,129); // Białystok
    g.addEdge(16,10,295); // Warszawa
    g.addEdge(16,15,196); // Olsztyn

    // Szczecin
    g.addEdge(17,13,111); // Gorzów Wlkp
    g.addEdge(17,18,161); // Koszalin

    // Koszalin
    g.addEdge(18,17,161); // Szczecin
    g.addEdge(18,14,140); // Pila
    g.addEdge(18,19,193); // Gdansk

    // Gdansk
    g.addEdge(19,15,168); // Olsztyn
    g.addEdge(19,12,170); // Torun
    g.addEdge(19,14,223); // Pila
    g.addEdge(19,18,193); // Koszalin
}
void ShortestPathDijkstra::clearGraph(){
    for (int i = 0; i < 0; ++i) {

    }
}

void ShortestPathDijkstra::startTimeTest() {
    prepareTestData();
    std::fstream fs = prepareFileWriter();
    for (int j = 0; j < iterations; ++j) {
        auto start = std::chrono::high_resolution_clock::now();
        g.shortestPath(0);
        g.shortestPath(1);
        g.shortestPath(2);
        g.shortestPath(3);
        g.shortestPath(4);
        g.shortestPath(5);
        g.shortestPath(6);
        g.shortestPath(7);
        g.shortestPath(8);
        g.shortestPath(9);
        g.shortestPath(10);
        g.shortestPath(11);
        g.shortestPath(12);
        g.shortestPath(13);
        g.shortestPath(14);
        g.shortestPath(15);
        g.shortestPath(16);
        g.shortestPath(17);
        g.shortestPath(18);
        g.shortestPath(19);

        diff = std::chrono::high_resolution_clock::now() - start;
        fs << std::setprecision(0) << std::fixed << diff.count() * 1000000000 << std::endl;
        std::cout << j << std::endl;
    }
    fs.close();
}

File: MEMORY_TEST/JAVA/DIJKSTRA_MEM/src/main/ShortestPathDijkstra.java
package main;

import main.dijkstra.Dijkstra;
import main.dijkstra.Graph;
import main.dijkstra.Node;


import java.io.PrintWriter;

public class ShortestPathDijkstra extends Algorithm {

    Graph graph;
    Node Krakow = new Node("Kraków");
    Node Rzeszów = new Node("Rzeszów");
    Node Katowice = new Node("Katowice");

    Node Wroclaw = new Node("Wrocław");
    Node Czestochowa = new Node("Czestochowa");
    Node Kielce = new Node("Kielce");
    Node Lublin = new Node("Lublin");

    Node ZielonaGora = new Node("Zielona Góra");
    Node Poznan = new Node("Poznań");
    Node Lodz = new Node("Lódź");
    Node Warszawa = new Node("Warszawa");

    Node Bialystok = new Node("Białystok");
    Node Torun = new Node("Toruń");
    Node GorzowWlkp = new Node("Gorzów Wielkopolski");
    Node Pila = new Node("Piła");

    Node Olsztyn = new Node("Olsztyn");
    Node Suwalki = new Node("Suwałki");
    Node Szczecin = new Node("Szczecin");
    Node Koszalin = new Node("Koszalin");
    Node Gdansk = new Node("Gdańsk");


    @Override
    public void prepareTestData() {
        graph = getClearGraph();
    }

    public Graph getClearGraph() {
        Graph g = new Graph();
//        Node Krakow = new Node("Kraków");
//        Node Rzeszów = new Node("Rzeszów");
//        Node Katowice = new Node("Katowice");
//
//        Node Wroclaw = new Node("Wrocław");
//        Node Czestochowa = new Node("Czestochowa");
//        Node Kielce = new Node("Kielce");
//        Node Lublin = new Node("Lublin");
//
//        Node ZielonaGora = new Node("Zielona Góra");
//        Node Poznan = new Node("Poznań");
//        Node Lodz = new Node("Lódź");
//        Node Warszawa = new Node("Warszawa");
//
//        Node Bialystok = new Node("Białystok");
//        Node Torun = new Node("Toruń");
//        Node GorzowWlkp = new Node("Gorzów Wielkopolski");
//        Node Pila = new Node("Piła");
//
//        Node Olsztyn = new Node("Olsztyn");
//        Node Suwalki = new Node("Suwałki");
//        Node Szczecin = new Node("Szczecin");
//        Node Koszalin = new Node("Koszalin");
//        Node Gdansk = new Node("Gdańsk");

        //Kraków
        Krakow.addDestination(Rzeszów,168);
        Krakow.addDestination(Kielce,116);
        Krakow.addDestination(Katowice,81);

        // Rzeszów
        Rzeszów.addDestination(Krakow,168);
        Rzeszów.addDestination(Kielce,157);
        Rzeszów.addDestination(Lublin,175);

        //Katowice
        Katowice.addDestination(Krakow,81);
        Katowice.addDestination(Czestochowa,75);
        Katowice.addDestination(Wroclaw,199);

        //Wrocław
        Wroclaw.addDestination(Katowice,199);
        Wroclaw.addDestination(Czestochowa,208);
        Wroclaw.addDestination(Lodz,222);
        Wroclaw.addDestination(Poznan,186);
        Wroclaw.addDestination(ZielonaGora,210);

        //Czestochowa
        Czestochowa.addDestination(Katowice,75);
        Czestochowa.addDestination(Wroclaw,208);
        Czestochowa.addDestination(Kielce,128);
        Czestochowa.addDestination(Lodz,126);

        //Kielce
        Kielce.addDestination(Czestochowa,128);
        Kielce.addDestination(Krakow,116);
        Kielce.addDestination(Rzeszów,157);
        Kielce.addDestination(Lublin,184);
        Kielce.addDestination(Warszawa,180);
        Kielce.addDestination(Lodz,159);

        //Lublin
        Lublin.addDestination(Rzeszów,175);
        Lublin.addDestination(Kielce,184);
        Lublin.addDestination(Warszawa,170);
        Lublin.addDestination(Bialystok,247);

        // Zielona Góra
        ZielonaGora.addDestination(Wroclaw, 210);
        ZielonaGora.addDestination(Poznan, 151);
        ZielonaGora.addDestination(GorzowWlkp, 112);

        //Poznan
        Poznan.addDestination(ZielonaGora,151);
        Poznan.addDestination(Wroclaw,186);
        Poznan.addDestination(Lodz,205);
        Poznan.addDestination(GorzowWlkp, 162);
        Poznan.addDestination(Torun, 167);

        //Lódź
        Lodz.addDestination(Wroclaw, 222);
        Lodz.addDestination(Czestochowa, 126);
        Lodz.addDestination(Kielce, 159);
        Lodz.addDestination(Warszawa, 136);
        Lodz.addDestination(Olsztyn, 297);
        Lodz.addDestination(Torun, 181);

        //Warszawa
        Warszawa.addDestination(Lublin,170);
        Warszawa.addDestination(Kielce,180);
        Warszawa.addDestination(Lodz,136);
        Warszawa.addDestination(Bialystok,203);
        Warszawa.addDestination(Suwalki,295);
        Warszawa.addDestination(Olsztyn,216);

        //Białystok
        Bialystok.addDestination(Lublin,247);
        Bialystok.addDestination(Warszawa,203);
        Bialystok.addDestination(Olsztyn,226);
        Bialystok.addDestination(Suwalki,129);

        //Toruń
        Torun.addDestination(Lodz,181);
        Torun.addDestination(Poznan,167);
        Torun.addDestination(Olsztyn, 176);
        Torun.addDestination(Gdansk, 170);
        Torun.addDestination(Pila, 143);

        //Gorzów Wlkp
        GorzowWlkp.addDestination(ZielonaGora,112);
        GorzowWlkp.addDestination(Poznan,162);
        GorzowWlkp.addDestination(Szczecin,111);
        GorzowWlkp.addDestination(Pila,121);

        //Piła
        Pila.addDestination(Torun,143);
        Pila.addDestination(GorzowWlkp,121);
        Pila.addDestination(Gdansk,223);
        Pila.addDestination(Koszalin,140);

        //Olsztyn
        Olsztyn.addDestination(Suwalki,196);
        Olsztyn.addDestination(Bialystok,226);
        Olsztyn.addDestination(Warszawa,216);
        Olsztyn.addDestination(Lodz,297);
        Olsztyn.addDestination(Torun,176);
        Olsztyn.addDestination(Gdansk,168);

        //Suwałki
        Suwalki.addDestination(Bialystok,129);
        Suwalki.addDestination(Warszawa,295);
        Suwalki.addDestination(Olsztyn,196);

        //Szczecin
        Szczecin.addDestination(GorzowWlkp,111);
        Szczecin.addDestination(Koszalin, 161);

        //Koszalin
        Koszalin.addDestination(Szczecin,161);
        Koszalin.addDestination(Pila,140);
        Koszalin.addDestination(Gdansk,193);

        //Gdańsk
        Gdansk.addDestination(Olsztyn,168);
        Gdansk.addDestination(Torun,170);
        Gdansk.addDestination(Pila,223);
        Gdansk.addDestination(Koszalin,193);

        g.addNode(Rzeszów);
        g.addNode(Krakow);
        g.addNode(Katowice);
        g.addNode(Wroclaw);
        g.addNode(Czestochowa);
        g.addNode(Kielce);
        g.addNode(Lublin);
        g.addNode(ZielonaGora);
        g.addNode(Poznan);
        g.addNode(Lodz);
        g.addNode(Warszawa);
        g.addNode(Bialystok);
        g.addNode(Torun);
        g.addNode(GorzowWlkp);
        g.addNode(Pila);
        g.addNode(Olsztyn);
        g.addNode(Suwalki);
        g.addNode(Szczecin);
        g.addNode(Koszalin);
        g.addNode(Gdansk);

        return g;
    }

    private void clearGraph() {
        for (Node it : graph.getNodes()
                ) {
            //System.out.println(it.getName() + " : " + it.getDistance());
            it.setDistance(2147483647);
        }
    }

//    private void clearGraph(Graph g) {
//        for (Node it : g.getNodes()
//                ) {
//            System.out.println(it.getName() + " : " + it.getDistance());
//            it.setDistance(2147483647);
//        }
//    }

//    public void clearAllGraphs() {
//        for (int i = 0; i < graphs.length; i++) {
//            clearGraph(graphs[i]);
//        }
//    }

    public ShortestPathDijkstra(String fileName, int iterations) {
        super(fileName, iterations);
    }

    @Override
    public void startTimeTest() {
        //PrintWriter pw = this.prepareFileWriter();
        prepareTestData();
        System.out.println("\t Data mem: " + (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()));
        for (int i = 0; i < this.getIterations(); i++) {
            timeDiff = System.nanoTime();

            Dijkstra.calculateShortestPathFromSource(graph, Krakow);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Rzeszów);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Katowice);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Wroclaw);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Czestochowa);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Kielce);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Lublin);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, ZielonaGora);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Poznan);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Lodz);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Warszawa);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Bialystok);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Torun);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, GorzowWlkp);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Pila);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Olsztyn);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Suwalki);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Szczecin);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Koszalin);
            clearGraph();
            Dijkstra.calculateShortestPathFromSource(graph, Gdansk);
            clearGraph();

            //pw.println( System.nanoTime() - timeDiff );
            //System.out.println(i);
        }

        //pw.close();
    }

    public static void main(String[] args) {
        ShortestPathDijkstra spd = new ShortestPathDijkstra("dijkstra_test",200000);
        spd.startTimeTest();
    }

}
